Bottom: d386bc5cbc520954e64a946513b18a251407c89e
Top:    0fd7db09ae99eeb9b439c2151b4b70fed0b63c9c
Author: Seva Gribov <seva.gribov@aerq.com>
Date:   2021-02-11 15:09:42 +0000

Spawn a child terminal


---

diff --git a/config.def.h b/config.def.h
index 93cbcc0..30978a6 100644
--- a/config.def.h
+++ b/config.def.h
@@ -201,6 +201,7 @@ static Shortcut shortcuts[] = {
 	{ TERMMOD,              XK_Num_Lock,    numlock,        {.i =  0} },
 	{ ShiftMask,            XK_Page_Up,     kscrollup,      {.i = -1} },
 	{ ShiftMask,            XK_Page_Down,   kscrolldown,    {.i = -1} },
+	{ TERMMOD,              XK_Return,      spawn,          {.i =  0} },
 };
 
 /*
diff --git a/st.c b/st.c
index 8d0e0ad..15cfc0f 100644
--- a/st.c
+++ b/st.c
@@ -1068,6 +1068,29 @@ tswapscreen(void)
 	tfulldirt();
 }
 
+void
+spawn(const Arg *dummy)
+{
+    ssize_t sz;
+    char shcwd[40];
+
+    snprintf(shcwd, LEN(shcwd), "/proc/%d/cwd", pid);
+    sz = readlink(shcwd, shcwd, LEN(shcwd));
+    if (sz == -1) {
+        perror("readlink");
+        return;
+    }
+    shcwd[sz] = '\0';
+
+    if (!fork()) {
+        if (chdir(shcwd) == -1) {
+            perror("chdir");
+        }
+        execl("/proc/self/exe", "st", NULL);
+        _exit(1);
+    }
+}
+
 void
 kscrolldown(const Arg* a)
 {
diff --git a/st.h b/st.h
index f44e1d3..c5868b3 100644
--- a/st.h
+++ b/st.h
@@ -81,6 +81,7 @@ void die(const char *, ...);
 void redraw(void);
 void draw(void);
 
+void spawn(const Arg *);
 void kscrolldown(const Arg *);
 void kscrollup(const Arg *);
 void printscreen(const Arg *);
